pagegraph/src/graph_algos.rs

```diff
@@ -433,32 +433,30 @@ impl PageGraph {
     }
 
     /// Get a collection of all Resource nodes whose requests match a set of adblock filter patterns.
-    /// Optionally, only match on exception patterns
-    pub fn resources_matching_filters(&self, patterns: Vec<String>, only_exceptions: bool) -> Vec<(NodeId, &Node)> {
+    pub fn resources_matching_filters(&self, graph: &PageGraph, patterns: Vec<String>) -> Vec<(String, String, Vec<String>, Vec<(usize, String, Option<String>,Option<String>)>)> {
         let source_url = self.root_url();
 
+        let mut matching_requests : Vec<(String, String, Vec<String>, Vec<(usize, String, Option<String>,Option<String>)>)> = vec![];
```

Comment:
Yes agreed! I was going to ask you if you have recommendations (but slipped my mind before creating the PR): I tried creating a struct but got the error that structs are not allowed in impls. Would `types.rs` be an okay place?

Reply:
I created a new commit showing what it might look like with the structs in `types.rs`. If you have suggestions for how to do it in a more Rust-y way I'm all ears 😅 I was trying to keep it performant and not have to call `check_network_urls_with_hostnames_subset()` more than once and carry the exception+blocking filter patterns to the calling `main()` function.

Reply:
> Yes agreed! I was going to ask you if you have recommendations (but slipped my mind before creating the PR): I tried creating a struct but got the error that structs are not allowed in impls. Would `types.rs` be an okay place? My hesitation was that file seemed more focused on PageGraph specific types

I think `types.rs` is a totally acceptable place for those definitions. You could also put them in `graph_algos.rs`, but you have to make sure it's outside of the `impl Pagegraph {...}` block or you'll get the error message you got above. I lean slightly towards the latter, as I think it's useful to distinguish between "core" graph datatypes and those that are only used for particular queries.

Otherwise, I think this looks pretty good!

Reply:
The latter makes more sense to me too, I did that.

Reply:
Should I merge it in?